src/java/org/apache/cassandra/net/FrameDecoder.java

```diff
@@ -263,9 +263,10 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws IOExceptio
     {
         if (msg instanceof BufferPoolAllocator.Wrapped)
         {
-            ByteBuffer buf = ((BufferPoolAllocator.Wrapped) msg).adopt();
+            BufferPoolAllocator.Wrapped wrapped = (BufferPoolAllocator.Wrapped) msg;
             // netty will probably have mis-predicted the space needed
-            allocator.putUnusedPortion(buf);
+            allocator.putUnusedPortion(wrapped);
+            ByteBuffer buf = ((BufferPoolAllocator.Wrapped) msg).adopt();
             channelRead(ShareableBytes.wrap(buf));
```

Comment:
@grighetto if the `buf` is not from buffer pool, we should not put it back to buffer pool when `ShareableBytes` is released; otherwise the metrics in `BufferPool` will be off.

how about returning a boolean from `allocator.putUnusedPortion(wrapped);` to indicate if it's pooled buffer, and then pass the boolean to `ShareableBytes`

Reply:
if we guarantee this is fixed-buffer use case (won't resize), we don't have to have to worry about that